File: scripts/python_include/flexio.py
import sys
import os
import random
import string
import json
import csv
import datetime
import pprint
import zmq
import base64
from zmq.utils.monitor import recv_monitor_message
from types import ModuleType



class CallProxy(object):
    def __init__(self, monitor=False):
        self.context = zmq.Context()
        self.socket = self.context.socket(zmq.REQ)
        self.socket.connect(os.environ['FLEXIO_RUNTIME_SERVER'])
        self.socket.setsockopt(zmq.RCVTIMEO, 250)
        if monitor:
            self.monitor = self.socket.get_monitor_socket(zmq.EVENT_CLOSED)
        else:
            self.monitor = None

    def invoke(self, method, params = []):

        call_id = ''.join(random.SystemRandom().choice(string.ascii_letters + string.digits) for _ in range(20))

        payload = {
            "version": 1,
            "access_key": os.environ['FLEXIO_RUNTIME_KEY'],
            "method": method,
            "params": params,
            "id": call_id
        }

        moniker = '~' + call_id + '/'
        self.convert_binary_to_base64(payload['params'], moniker)

        self.socket.send(json.dumps(payload).encode())

        while True:
            try:
                resobj = self.socket.recv_json()
                return self.convert_base64_to_binary(resobj['result'], moniker)
            except zmq.ZMQError:
                pass

            if self.monitor:
                try:
                    evt = recv_monitor_message(self.monitor, zmq.NOBLOCK)
                    if evt['event'] == zmq.EVENT_CLOSED:
                        print("Connection broken")
                        sys.exit(1)
                except zmq.ZMQError:
                    pass

    def close(self):
        self.socket.close()
        self.socket = None

    def convert_binary_to_base64(self, var, moniker):
        if isinstance(var, list):
            for key, value in enumerate(var):
                var[key] = self.convert_binary_to_base64(var[key], moniker)
        elif isinstance(var, dict):
            for key, value in var.items():
                var[key] = self.convert_binary_to_base64(var[key], moniker)
        elif isinstance(var, (bytes, bytearray)):
            return moniker + 'bin.b64:' + base64.b64encode(var).decode()
        elif isinstance(var, object) and getattr(var, '_handle', None) is not None:
            return moniker + 'stream:' + str(var._handle)
        return var

    def convert_base64_to_binary(self, var, moniker):
        moniker_len = len(moniker)
        if isinstance(var, list):
            for key, value in enumerate(var):
                var[key] = self.convert_base64_to_binary(var[key], moniker)
        elif isinstance(var, dict):
            for key, value in var.items():
                var[key] = self.convert_base64_to_binary(var[key], moniker)
        elif isinstance(var, str) and var[:moniker_len+8] == moniker + 'bin.b64:':
            return base64.b64decode(var[moniker_len+8:].encode())
        return var



proxy = CallProxy()




class EnvVars(object):

    def __init__(self):
        self.fx_inited = False
        self.fx_env = {}

    def initialize(self):
        self.fx_env = proxy.invoke('getEnv', [])
        self.fx_inited = True

    def __getitem__(self, key):
        if not self.fx_inited:
            self.initialize()
        return self.fx_env[key]

    def __getattr__(self, key):
        if not self.fx_inited:
            self.initialize()
        if key in self.fx_env:
            return self.fx_env[key]
        else:
            return None

    def __setitem__(self, key, value):
        if not self.fx_inited:
            self.initialize()
        if proxy.invoke('setEnvValue', [key,value]) is True:
            self.fx_env[key] = value

    def __setattr__(self, key, value):
        if key[:3] == 'fx_':
            return super().__setattr__(key, value)
        if not self.fx_inited:
            self.initialize()
        if proxy.invoke('setEnvValue', [key,value]) is True:
            self.fx_env[key] = value

    def __iter__(self):
        if not self.fx_inited:
            self.initialize()
        return iter(self.fx_env)

    def items(self):
        if not self.fx_inited:
            self.initialize()
        return self.fx_env.items()

    def keys(self):
        if not self.fx_inited:
            self.initialize()
        return self.fx_env.keys()

    def values(self):
        if not self.fx_inited:
            self.initialize()
        return self.fx_env.values()

env_vars_obj = EnvVars()





class Connection(object):
    def __init__(self, info):
        self.eid = info['eid']
        self.name = info['name']
        self.description = info['description']

    def get_credentials(self):
        return proxy.invoke('getConnectionCredentials', [self.eid])

    def get_access_token(self):
        return proxy.invoke('getConnectionAccessToken', [self.eid])



class ContextConnections(object):

    def __init__(self):
        self.fx_inited = False
        self.fx_connection_list = []
        self.fx_connection_map = {}

    def initialize(self):
        connections = proxy.invoke('getConnections', [])
        for connection in connections:
            connobj = Connection(connection)
            self.fx_connection_map[connobj.eid] = connobj
            self.fx_connection_map[connobj.name] = connobj
            self.fx_connection_list.append(connobj)
        connections = proxy.invoke('getLocalConnections', [])
        for connection in connections:
            connobj = Connection(connection)
            self.fx_connection_map[connobj.name] = connobj
        self.fx_inited = True

    def __getitem__(self, key):
        if not self.fx_inited:
            self.initialize()
        return self.fx_connection_map[key]

    def __getattr__(self, key):
        if not self.fx_inited:
            self.initialize()
        if key in self.fx_connection_map:
            return self.fx_connection_map[key]
        else:
            return None

    def __iter__(self):
        if not self.fx_inited:
            self.initialize()
        return iter(self.fx_connection_list)

    def items(self):
        if not self.fx_inited:
            self.initialize()
        return self.fx_connection_list.items()

    def keys(self):
        if not self.fx_inited:
            self.initialize()
        return self.fx_connection_list.keys()

    def values(self):
        if not self.fx_inited:
            self.initialize()
        return self.fx_connection_list.values()

context_connections_obj = ContextConnections()



class Stream(object):

    # used for setting fetch style
    Dictionary = 1
    Tuple = 2

    def __init__(self, info):
        if info:
            self._name = info['name']
            self._content_type = info['content_type']
            self._size = info['size']
            self._handle = info['handle']
            self._structure = None
            self._dict = False
            self._fetch_style = self.Dictionary
            self._casts = None
            self._casting = True
            self._need_commit = False
        else:
            self._name = ''
            self._content_type = 'application/octet-stream'
            self._size = 0
            self._handle = -1
            self._structure = None
            self._dict = False
            self._fetch_style = self.Dictionary
            self._casts = None
            self._casting = True
            self._need_commit = False

    def __del__(self):
        if self._handle != 0 and self._need_commit:
            self.close()

    @property
    def name(self):
        return self._name

    @name.setter
    def name(self, value):
        self._name = value
        proxy.invoke('setStreamInfo', [self._handle, {'name':value}])

    @property
    def content_type(self):
        return self._content_type

    @content_type.setter
    def content_type(self, value):
        self._content_type = value
        proxy.invoke('setStreamInfo', [self._handle, {'content_type':value}])

    @property
    def size(self):
        return self._size

    @property
    def structure(self):
        if not self._structure:
            self._structure = proxy.invoke('getInputStreamStructure', [self._handle])
        return self._structure

    @property
    def fetch_style(self):
        return self._fetch_style

    @fetch_style.setter
    def fetch_style(self, value):
        self._casts = None
        if type({}) == value:
            self._fetch_style = self.Dictionary
        elif type(()) == value:
            self._fetch_style = self.Tuple
        else:
            raise ValueError("fetch style must be set to dict or tuple")

    @property
    def casting(self):
        return self._casting

    @casting.setter
    def casting(self, value):
        self._casting = value

    def create(self, name=None, content_type='text/plain', structure=None):
        properties = { 'content_type': content_type }
        if name:
            properties['name'] = name
        if structure:
            properties['structure'] = structure
        if proxy.invoke('managedCreate', [self._handle, properties]) is False:
            return False
        else:
            return self

    def read(self, length=None):
        if length is None:
            return self.readall()
        data = proxy.invoke('read', [self._handle, length])
        if data is False:
            return b''
        return data

    def readline(self):
        if self._fetch_style == self.Tuple:
            row = proxy.invoke('readline', [self._handle, False])
            if row is False:
                return None
            if self._casting:
                self.type_casts(row)
            return tuple(row)
        else:
            row = proxy.invoke('readline', [self._handle, True])
            if row is False:
                return None
            if self._casting:
                self.type_casts(row)
            return row

    def readall(self):
        buf = b''
        while True:
            chunk = proxy.invoke('read', [self._handle, 4096])
            if chunk is False:
                break
            buf += chunk
        return buf

    def __iter__(self):
        return self

    def __next__(self):
        row = self.readline()
        if row is None:
            raise StopIteration
        return row

    def type_casts(self, row):
        if not isinstance(row, dict):
            return row
        if self._casts is None:
            self._casts = {}
            structure = self.structure
            idx = 0
            for col in self.structure:
                if self._fetch_style == self.Tuple:
                    key = idx
                else:
                    key = col['name']
                if col['type'] == 'numeric':
                    self._casts[key] = float
                elif col['type'] == 'integer':
                    self._casts[key] = int
                elif col['type'] == 'date':
                    self._casts[key] = lambda s: datetime.datetime.strptime(s, '%Y-%m-%d').date()
                elif col['type'] == 'datetime':
                    self._casts[key] = lambda s: datetime.datetime.strptime(s, '%Y-%m-%d')
                idx = idx + 1
        for key,func in self._casts.items():
            try:
                row[key] = func(row[key])
            except IndexError:
                continue
            except KeyError:
                continue

    def write(self, data):
        proxy.invoke('write', [self._handle, data])

    def insert_row(self, row):
        proxy.invoke('insertRow', [self._handle, row])

    def insert_rows(self, rows):
        proxy.invoke('insertRows', [self._handle, rows])

    def close(self):
        if self._handle != 0 and self._need_commit:
            proxy.invoke('fsCommit', [self._handle])
            self._need_commit = False



class ContextFs(object):

    def __init__(self):
        pass

    def create(self, path, connection=''):
        info = proxy.invoke('fsCreate', [path, connection])
        stream = Stream(info)
        stream._need_commit = True
        return

    def create_tempfile(self):
        info = proxy.invoke('fsCreateTempFile', [])
        stream = Stream(info)
        stream._need_commit = True
        return stream

    def open(self, path, mode='r', connection=''):
        info = proxy.invoke('fsOpen', [mode, path, connection])
        stream = Stream(info)
        if mode == 'w' or mode == 'w+' or mode == 'a' or mode == 'a+':
            stream._need_commit = True
        return stream

    def read(self, path, connection=''):
        info = proxy.invoke('fsOpen', ['r', path, connection])
        stream = Stream(info)
        return stream.readall()

    def write(self, path, data='', connection=''):

        info = proxy.invoke('fsOpen', ['w', path, connection])
        handle = info['handle']

        idx = 0
        buf = data
        while len(buf) > 0:
            chunk = buf[:4096]
            buf = buf[4096:]
            proxy.invoke('write', [handle, chunk])
            idx = idx + 1

        handle = proxy.invoke('fsCommit', [handle])


    def list(self, path, connection=''):
        return proxy.invoke('fsList', [path, connection])

    def exists(self, path, connection=''):
        return proxy.invoke('fsExists', [path, connection])

    def remove(self, path, connection=''):
        return proxy.invoke('fsRemove', [path, connection])



class ContextEmail(object):

    def __init__(self):
        pass

    def send(self, *args, **kwargs):
        info = proxy.invoke('invoke_service', ['email.send', args[0] if len(args) == 1 else kwargs])



class ContextKeyValue(object):
    def __init__(self):
        pass

    def set(self, k, v):
        return proxy.invoke('kvSet', [k, v])

    def get(self, k):
        return proxy.invoke('kvGet', [k])

    def incr(self, k, v=1):
        return proxy.invoke('kvIncr', [k,v])

    def decr(self, k, v=1):
        return proxy.invoke('kvDecr', [k,v])



class ContextIndex(object):

    def __init__(self):
        pass

    def create(self, name, params={}):
        return proxy.invoke('indexCreate', [name, json.dumps(params)])

    def remove(self, name):
        return proxy.invoke('indexRemove', [name])

    def clear(self, name):
        return proxy.invoke('indexClear', [name])

    def insert(self, name, params):
        return proxy.invoke('indexInsert', [name, json.dumps(params)])

    def exists(self, name):
        return proxy.invoke('indexExists', [name])

    def rename(self, old_name, new_name):
        return proxy.invoke('indexRename', [old_name, new_name])


class Context(object):
    def __init__(self):
        self._input = None
        self._output = None
        self._query = None
        self._form = None
        self._files = None
        self.email = ContextEmail()
        self.fs = ContextFs()
        self.kv = ContextKeyValue()
        self.index = ContextIndex()

    @property
    def input(self):
        if self._input is None:
            stdin_stream_info  = proxy.invoke('getStreamInfo', [0])
            self._input = Stream(stdin_stream_info)
        return self._input

    @property
    def output(self):
        if self._output is None:
            stdout_stream_info  = proxy.invoke('getStreamInfo', [1])
            self._output = Stream(stdout_stream_info)
        return self._output

    @property
    def query(self):
        if self._query is None:
            self._query = proxy.invoke('getQueryParameters', [])
        return self._query

    @property
    def form(self):
        if self._form is None:
            self._form = proxy.invoke('getFormParameters', [])
        return self._form

    @property
    def files(self):
        if self._files is None:
            self._files = {}
            fileinfos = proxy.invoke('getFilesParameters', [])
            for fileinfo in fileinfos:
                self.files[fileinfo['tag']] = Stream(fileinfo)
        return self._files

    @property
    def vars(self):
        return env_vars_obj

    @property
    def connections(self):
        return context_connections_obj

    def json(self, obj):
        self.output.content_type = "application/json"
        self.output.write(json.dumps(obj))
        proxy.close()
        sys.exit()

    def status(self, code):
        fileinfos = proxy.invoke('status', [code])

    def end(self, content=None, content_type=None):
        if content_type is not None:
            self.output.content_type = content_type
        if content is not None:
            self.output.write(content)
        proxy.close()
        sys.exit()



def run(handler):

    context = Context()

    if isinstance(handler, ModuleType):
        func = getattr(handler, 'flex_handler', None)
        if callable(func):
            func(context)
            return
        func = getattr(handler, 'flexio_handler', None)
        if callable(func):
            func(context)
            return
    else:
        handler(context)

File: scripts/nodejs_include/flexio.js
"use strict"

var utf8 = require('utf8')
var fs = require('fs')
var Flexio = require('flexio-sdk-js')
var zmq = require('zeromq')

var proxy, context
var input = null
var output = null


class CallProxy {


    constructor(config) {

        var pThis = this

        this.server = process.env.hasOwnProperty('FLEXIO_RUNTIME_SERVER') ? ''+process.env.FLEXIO_RUNTIME_SERVER : ''
        this.access_key = process.env.hasOwnProperty('FLEXIO_RUNTIME_KEY') ? ''+process.env.FLEXIO_RUNTIME_KEY : ''
        this.requester = null
        this.calls = {}

        if (config !== null && typeof config == 'object') {
            if (config.hasOwnProperty('server')) {
                this.server = config.server
            }
            if (config.hasOwnProperty('access_key')) {
                this.access_key = config.access_key
            }
        }


        if (this.server.length > 0)
        {
            this.requester = zmq.socket('req')
            this.requester.connect(this.server)

            this.requester.on("message", function(msg) {
                msg = msg.toString()
                //console.log("RECEIVED MESSAGE " + msg)
                pThis.onMessage(msg)
            })

            process.on('SIGINT', function() {
                pThis.requester.close();
            });
        }


    }


    close() {
        this.requester.close()
        this.requester = null
    }

    onMessage(reply) {
        
        var result = undefined, call_id = ''

        try
        {
            reply = JSON.parse(reply)
            if (!reply.hasOwnProperty('id'))
                throw "Call response is missing id"
            if (reply.hasOwnProperty('result'))
                result = reply.result
            call_id = reply.id
            var moniker = '~' + call_id + '/bin.b64:'
            result = this.convertBase64ToBinary(result, moniker)
        }
        catch (e)
        {
        }

        if (this.calls.hasOwnProperty(call_id))
        {
            var callback = this.calls[call_id]
            delete this.calls[call_id]
            callback(result)
        }
    }

    invokeAsync(method, params) {

        var pThis = this

        var call_id = [...Array(20)].map(i=>(~~(Math.random()*36)).toString(36)).join('')
        var moniker = '~' + call_id + '/bin.b64:'

        var payload = {
            "version": 1,
            "access_key": this.access_key,
            "method": method,
            "params": this.convertBinaryToBase64(params, moniker),
            "id": call_id
        }

        return new Promise(function(resolve, reject) {
            pThis.calls[call_id] = resolve
            var msg = JSON.stringify(payload)
            //console.log("SENDING " + msg)
            pThis.requester.send(msg)
        })
    }


    invokeSync(method, params) {

        var done = false
        var result = undefined

        this.invokeAsync(method, params).then(function(r) {
            //console.log("Function returned")
            result = r
            done = true
        }).catch(function(err) {

        })

        //console.log("waiting")
        require('deasync').loopWhile(function(){return !done})
        //console.log("done waiting")

        return result
    }


    convertBinaryToBase64(obj, moniker)
    {
        var res, k, len

        if (Array.isArray(obj)) {
            res = []
            len = obj.length
            for (k = 0; k < len; ++k) {
                res.push(this.convertBinaryToBase64(obj[k], moniker))
            }
            return res
        } else if (Object.prototype.toString.call(obj) === '[object Object]' /* if is plain object */) {
            res = {}
            for (k in obj) {
                if (obj.hasOwnProperty(k)) {
                    res[k] = this.convertBinaryToBase64(obj[k], moniker)
                }
            }
            return res
        } else if (obj instanceof Buffer) {
            return moniker + obj.toString('base64')
        } else {
            return obj
        }
          
    }
    
    convertBase64ToBinary(obj, moniker)
    {
        var res, k, len

        if (Array.isArray(obj)) {
            res = []
            len = obj.length
            for (k = 0; k < len; ++k) {
                res.push(this.convertBase64ToBinary(obj[k], moniker))
            }
            return res
        } else if (Object.prototype.toString.call(obj) === '[object Object]' /* if is plain object */) {
            res = {}
            for (k in obj) {
                if (obj.hasOwnProperty(k)) {
                    res[k] = this.convertBase64ToBinary(obj[k], moniker)
                }
            }
            return res
        } else if ((typeof obj === 'string' || obj instanceof String) && obj.substr(0, moniker.length) == moniker) {
            return Buffer.from(obj.substr(moniker.length), 'base64')
        } else {
            return obj
        }
    }
}



class InputEnv {
}

class OutputEnv {
}

var inputEnv = new InputEnv()
var outputEnv = new OutputEnv()




class Input {

    constructor(info) {
        // used for setting fetch style
        this.FETCH_OBJECT = 1
        this.FETCH_ARRAY = 2

        if (info) {
            this._name = info['name']
            this._contentType = info['content_type']
            this._isTable = (this.contentType == 'application/vnd.flexio.table') ? true:false
            this._size = info['size']
            this._handle = info['handle']
            this._structure = null
            this._dict = false
            this._fetchStyle = this.Dictionary
            this._casts = null
            this._casting = true
        } else {
            this._name = ''
            this._contentType = 'application/octet-stream'
            this._isTable = false
            this._size = 0
            this._handle = -1
            this._structure = null
            this._dict = false
            this._fetchStyle = this.Dictionary
            this._casts = null
            this._casting = true
        }
    }
    
    get env() {
        return inputEnv
    }

    get name() {
        return this._name
    }

    get contentType() {
        return this._contentType
    }

    get size() {
        return this._size
    }

    get structure() {
        if (!this._isTable) {
            return null
        }
        if (!this._structure) {
            this._structure = proxy.invokeSync('getInputStreamStructure', [this._handle])
        }
        return this._structure
    }
    
    get isTable() {
        return this._isTable
    }

    get fetchStyle() {
        return this._fetchStyle
    }

    set fetchStyle(value) {
        this._casts = null
        if (value == Object || value == this.FETCH_OBJECT) {
            this._fetchStyle = this.FETCH_OBJECT
        } else if (value == Array || value == this.FETCH_ARRAY) {
            this._fetchStyle = this.FETCH_ARRAY
        }
    }

    get casting() {
        return this._casting
    }

    set casting(value) {
        this._casting = value
    }

    read(length) {
        if (length === undefined || length === null) {
            return this.readAll()
        }
        if (this._isTable) {
            return null
        }
        var data = proxy.invokeSync('read', [this._handle, length])
        if (data === false || data === null) {
            return null
        }
        return data
    }

    readLine() {
        var row = proxy.invokeSync('readline', [this._handle, (this._fetchStyle == this.FETCH_ARRAY ? false : true)])
        if (row === false || row === null) {
            return null
        }
        if (this._casting) {
            this.typeCasts(row)
        }
        return row
    }
    
    forEach(callback) {
        var row
        while ((row = this.readLine()) !== null) {
            callback(row)
        }
    }
    
    readAll() {
        if (this._isTable) {
            var rows = [], row
            while (true) {
                row = this.readLine()
                if (row === null) {
                    break
                }
                rows.push(row)
            }
            return rows
        } else {
            var buf = '', chunk
            while (true) {
                chunk = this.read(4096)
                if (chunk === null) {
                    break
                }
                buf += chunk
            }
            return buf
        }
    }

    typeCasts(row) {
        var key
        if (this._casts === null) {
            this._casts = {}
            var structure = this.structure
            var i, idx = 0, colCount = structure.length, col
            for (i = 0; i < colCount; ++i) {
                col = structure[i]
                if (this._fetchStyle == this.FETCH_ARRAY) {
                    key = idx
                } else {
                    key = col['name']
                }
                if (col['type'] == 'numeric') 
                    this._casts[key] = function(val) { return parseFloat(val) }
                else if (col['type'] == 'integer')
                    this._casts[key] = function(val) { return parseInt(val) }
                else if (col['type'] == 'date')
                    this._casts[key] = function(val) { return new Date(val) }
                else if (col['type'] == 'datetime')
                    this._casts[key] = function(val) { return new Date(val) }
                idx = idx + 1
            }
        }

        for (key in this._casts) {
            if (this._casts.hasOwnProperty(key) && row.hasOwnProperty(key)) {
                row[key] = this._casts[key](row[key])
            }
        }
    }
}

/*
    def __iter__(self):
        return self

    def __next__(self):
        row = self.readline()
        if row is None:
            raise StopIteration
        return row


*/





class Output {


    constructor(info) {
        if (info) {
            this._name = info['name']
            this._contentType = info['content_type']
            this._size = info['size']
            this._handle = info['handle']
        } else {
            this._name = ''
            this._contentType = 'application/octet-stream'
            this._size = 0
            this._handle = -1
        }
    }
    
    get name() {
        return this._name
    }

    set name(value) {
        this._name = value
        proxy.invokeSync('setOutputStreamInfo', [this._handle, {'name':value}])
    }

    get contentType() {
        return this._contentType
    }

    set contentType(value) {
        this._contentType = value
        proxy.invokeSync('setOutputStreamInfo', [this._handle, {'content_type':value}])
    }

    get env() {
        return outputEnv
    }

//  create(structure)
//  create(name, structure)
//  create(name, contentType, structure)
    create(p1,p2,p3) {
        var name, structure, contentType, properties
        if (Array.isArray(p1)) {
            // handle this call format:  create(structure)
            structure = p1
            name = null
            contentType = 'application/vnd.flexio.table'
        }
        else if (Array.isArray(p2)) {
            // handle this call format:  create(name, structure)
            name = p1
            structure = p2
            contentType = 'application/vnd.flexio.table'
        }
        else {
            name = (p1 === undefined) ? null : p1
            structure = (p2 === undefined) ? null : p2
            contentType = (p3 === undefined) ? 'text/plain' : p3
        }
        properties = { 'content_type': contentType }
        if (name) {
            properties['name'] = name
        }
        if (structure) {
            properties['structure'] = structure
        }
        var res = proxy.invokeSync('managedCreate', [this._handle, properties])

        if (res === false) {
            return false
        } else {
            return this
        }
    }

    write(data) {
        this.typeCasts(data)
        proxy.invokeSync('write', [this._handle, data])
    }

    insertRow(row) {
        this.typeCasts(row)
        proxy.invokeSync('insertRow', [this._handle, row])
    }
    
    insertRows(rows) {
        for (var i = 0, cnt = rows.length; i < cnt; ++i) {
            this.typeCasts(rows[i])
        }
        proxy.invokeSync('insertRows', [this._handle, rows])
    }


    castValue(value) {
        if (value instanceof Date) {
            var y = value.getFullYear(), m = value.getMonth()+1, d = value.getDate()
            return '' + y + '-' + (m<10?'0':'') + m + '-' + (d<10?'0':'') + d
        } else {
            return value
        }
    }

    typeCasts(row) {
        if (row instanceof Buffer) {
        } else if (Array.isArray(row)) {
            for (var i = 0, cnt = row.length; i < cnt; ++i) {
                row[i] = this.castValue(row[i])
            }
        } else if (typeof row === 'object') {
            for (var key in row) {
                if (row.hasOwnProperty(key)) {
                    row[key] = this.castValue(row[key])
                }
            }
        }
    }
}


class Result {
    json(obj) {
        output.contentType = "application/json"
        output.write(JSON.stringify(obj))
        proxy.close()
    }

    end(content) {
        if (content !== null && content !== undefined) {
            output.write(content)
        }
        proxy.close()
    }
}


class Context {

    constructor() {
        var pThis = this
        this.res = new Result()
        this._query = null
        this._form = null
        this._vars = {}
        this._vars_inited = false

        this.pipe = {}
        for (var task_name in Flexio.task) {

            (function(task_name, task_func){
                if (Flexio.task.hasOwnProperty(task_name) && task_name != 'toCode') {
                  pThis.pipe[task_name] = function() {
                      proxy.invokeSync('runJob', [ JSON.stringify(task_func.apply(this, arguments)) ])
                      return pThis.pipe
                  }
                }
              })(task_name, Flexio.task[task_name])
        }


        this.vars = new Proxy(pThis._vars, {
            checkPopulate: function(target) {
                if (!pThis._vars_inited) {
                    var env = proxy.invokeSync('getEnv', [])
                    var keys = Object.keys(env)
                    for (var i = 0; i < keys.length; ++i) {
                        target[keys[i]] = env[keys[i]]
                    }
                    pThis._vars_inited = true
                }
            },
            get: function(target, prop, receiver) {
                this.checkPopulate(target)
                return pThis._vars.hasOwnProperty(prop) ? pThis._vars[prop] : null
            },
            set: function(target, prop, value) {
                this.checkPopulate(target)
                pThis._vars[prop] = value
                proxy.invokeSync('setEnvValue', [prop,value])
            },
            ownKeys: function(target) {
                this.checkPopulate(target)
                return Object.keys(target)
            }
        });
        
    }

    get query() {
        if (this._query === null) {
            this._query = proxy.invokeSync('getQueryParameters', [])
        }
        return this._query
    }

    get form() {
        if (this._form === null) {
            this._form = proxy.invokeSync('getFormParameters', [])
        }
        return this._form
    }

}





var inited = false
function checkModuleInit(callback) {
    if (inited) {
        callback()
        return
    } else {
        inited = true
        proxy.invokeAsync('getInputStreamInfo', ['_fxstdin_']).then((stdin_stream_info) => {
            input = new Input(stdin_stream_info)
            context.input = input
            proxy.invokeAsync('getOutputStreamInfo', ['_fxstdout_']).then((stdout_stream_info) => {
                output = new Output(stdout_stream_info)
                context.output = output
                inited = true
                callback()
            }).catch((err)=>{ console.log("Error initializing stdin stream"); process.exit(1); })
        }).catch((err)=>{ console.log("Error initializing stdout stream"); process.exit(1); })
    }
}


function isFunction(f) {
    return f && {}.toString.call(f) === '[object Function]';
}

function run(handler) {

    if (handler.hasOwnProperty('flex_handler')) {
        handler = handler.flex_handler
    }
    else if (handler.hasOwnProperty('flexio_handler')) {
        handler = handler.flexio_handler
    }

    if (isFunction(handler)) {
        proxy = new CallProxy()
        context = new Context()
        context.proxy = proxy
        
        checkModuleInit(function() {
            handler(context)
        })
    }
}



module.exports = {

    CallProxy,
    run

};
